## RAG DEVELOPMENT

In [34]:
from google.colab import drive
drive.mount('/gdrive')
%cd /gdrive

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).
/gdrive


In [35]:
cd MyDrive/RAG_projects/DS-RPC-01/

/gdrive/MyDrive/RAG_projects/DS-RPC-01


## Install OpenAI, and LangChain dependencies

In [3]:
!pip install langchain==0.3.10
!pip install langchain-openai==0.2.12
!pip install langchain-community==0.3.11
!pip install dill

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 3.8 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of langchain-core to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 24.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 438.5/438.5 kB 27.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.8/311.8 kB 22.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 56.7 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.0.2
    Uninstalling numpy-2.0.2:
      Successfully uninstalled numpy-2.0.2
  Attempting uninstall: langsmith
    Found existing installation: langsmith 0.3.45
    Uninstalling langsmith-0.3.45:
      Successfully uninstalled langsmith-0.3.45
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.3.65
    Uninstalling langchain-core-0.3.65:
  

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.7/50.7 kB 3.1 MB/s eta 0:00:00
  Using cached langchain_core-0.3.66-py3-none-any.whl.metadata (5.8 kB)
INFO: pip is looking at multiple versions of langchain-core to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 37.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 36.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.2/45.2 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 3.9 MB/s eta 0:00:00
  Attempting uninstall: langchain
    Found existing installation: langchain 0.3.10
    Uninstalling langchain-0.3.10:
      Successfully uninstalled langchain-0.3.10


In [4]:
# takes 2 - 5 mins to install on Colab
!pip install "unstructured[all-docs]==0.14.0"

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 29.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.2/42.2 kB 3.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 kB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.2/48.2 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 63.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.5/59.5 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 471.6/471.6 kB 29.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 590.6/590.6 kB 38.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.9/527.9 kB 33.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.8/101.8 kB 9.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.6/17.6 MB 60.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [3]:
import nltk
nltk.download('punkt_tab')
nltk.download('averaged_perceptron_tagger_eng')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!


True

## Install Chroma Vector DB and LangChain wrapper

In [6]:
!pip install langchain-chroma==0.1.4

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 5.3 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 628.3/628.3 kB 19.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 64.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 284.2/284.2 kB 16.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 68.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.6/101.6 kB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.8/65.8 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.7/55.7 kB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 196.2/196.2 kB 14.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.5/118.5 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 105.4/105.4 kB 7.1 MB/s eta 

## Enter Open AI API Key

In [4]:
# Enter Open AI API Key
from getpass import getpass

OPENAI_KEY = getpass('Enter Open AI API Key: ')

Enter Open AI API Key: ··········


In [5]:
# Setup Environment Variables
import os

os.environ['OPENAI_API_KEY'] = OPENAI_KEY

### Open AI Embedding Models

LangChain enables us to access Open AI embedding models which include the newest models: a smaller and highly efficient `text-embedding-3-small` model, and a larger and more powerful `text-embedding-3-large` model.

In [6]:
from langchain_openai import OpenAIEmbeddings

# details here: https://openai.com/blog/new-embedding-models-and-api-updates
openai_embed_model = OpenAIEmbeddings(model='text-embedding-3-small')

## Loading and Processing the Data

### Get the dataset

In [36]:
# All the processed data will be stored in processed_docs
# which going to ingest in to chroma DB
processed_docs = []

#### HR DATASET

In [37]:
import pandas as pd
# loading hr data
df = pd.read_csv('data/hr/hr_data.csv')
print('Shape of dataset',df.shape)
df.head(2)

Shape of dataset (100, 15)


employee_id       full_name            role department  \
0  FINEMP1000    Aadhya Patel   Sales Manager      Sales   
1  FINEMP1001  Isha Chowdhury  Credit Officer    Finance   

                          email   location date_of_birth date_of_joining  \
0    aadhya.patel@fintechco.com  Ahmedabad    1991-04-03      2018-11-20   
1  isha.chowdhury@fintechco.com       Pune    1995-09-21      2021-05-20   

   manager_id      salary  leave_balance  leaves_taken  attendance_pct  \
0  FINEMP1006  1332478.37             22            11           99.31   
1  FINEMP1005  1491158.23              8             3           85.15   

   performance_rating last_review_date  
0                   3       2024-05-21  
1                   5       2024-01-20

In [38]:
from langchain.docstore.document import Document

# Document loaders to load data from a source as a Document data type object
# Document has a piece of text content and associated metadata

for index, row in df.iterrows():
    content = (
        f"Employee {row['full_name']} (ID: {row['employee_id']}) works as a {row['role']} in the {row['department']} department. "
        f"Their email is {row['email']} and they are based in {row['location']}. "
        f"Born on {row['date_of_birth']}, they joined the company on {row['date_of_joining']} and report to manager ID {row['manager_id']}. "
        f"Their current salary is {row['salary']}, with {row['leave_balance']} leave days remaining out of {row['leaves_taken']} taken. "
        f"Attendance percentage is {row['attendance_pct']}% and their latest performance rating is {row['performance_rating']} "
        f"(last reviewed on {row['last_review_date']})."
    )
    metadata = row.to_dict()
    metadata["file_name"] = "hr_data.csv"
    metadata["page_no"] = index + 1
    metadata["access_level"] = "hr"

    # metadata is optional but recommended - core usage is during retrivel based on filter/condition, without metadata difficult process based on conditions
    processed_docs.append(Document(page_content=content, metadata=metadata))


In [39]:
print('Total Processed data count', len(processed_docs))
# sample data to view
processed_docs[0]

Total Processed data count 100


Document(metadata={'employee_id': 'FINEMP1000', 'full_name': 'Aadhya Patel', 'role': 'Sales Manager', 'department': 'Sales', 'email': 'aadhya.patel@fintechco.com', 'location': 'Ahmedabad', 'date_of_birth': '1991-04-03', 'date_of_joining': '2018-11-20', 'manager_id': 'FINEMP1006', 'salary': 1332478.37, 'leave_balance': 22, 'leaves_taken': 11, 'attendance_pct': 99.31, 'performance_rating': 3, 'last_review_date': '2024-05-21', 'file_name': 'hr_data.csv', 'page_no': 1, 'access_level': 'hr'}, page_content='Employee Aadhya Patel (ID: FINEMP1000) works as a Sales Manager in the Sales department. Their email is aadhya.patel@fintechco.com and they are based in Ahmedabad. Born on 1991-04-03, they joined the company on 2018-11-20 and report to manager ID FINEMP1006. Their current salary is 1332478.37, with 22 leave days remaining out of 11 taken. Attendance percentage is 99.31% and their latest performance rating is 3 (last reviewed on 2024-05-21).')

#### FINANCE DATASET

In [40]:
# The Finance dataset is in .md(markdown) format
# LangChain implements an UnstructuredMarkdownLoader object which uses the Unstructured library to load data from Markdown files
# Langchain uses the Unstructured.io(famous doc loader opensource) library under the hood.

from langchain_community.document_loaders import UnstructuredMarkdownLoader

In [41]:
import tiktoken

def count_tokens_in_file(file_path: str, model: str = "gpt-4") -> int:
    # Load tokenizer
    encoding = tiktoken.encoding_for_model(model)

    # Read content
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    # Tokenize and count
    tokens = encoding.encode(content)
    print(f"Total tokens in '{file_path}': {len(tokens)}")
    return None


In [42]:
count_tokens_in_file('data/finance/financial_summary.md')

Total tokens in 'data/finance/financial_summary.md': 1420


GPT-4o Mini supports a context window of up to 128,000 tokens, allowing for substantial input sizes. Additionally, it can generate up to 16,384 tokens in a single response. ([techtarget.com][1])

However, if you're using GPT-4o Mini through Azure OpenAI, be aware that some users have reported practical input token limits around 35,000 tokens, despite the documented 128,000-token context window. This discrepancy may be due to specific deployment configurations or subscription tiers. ([learn.microsoft.com][2])

In your case, with a 1420-token document, you're well within both the official and observed limits. Therefore, we're proceed to load the document without splitting(mode = single).

[1]: https://www.techtarget.com/whatis/feature/GPT-4o-explained-Everything-you-need-to-know?utm_source=chatgpt.com "GPT-4o explained: Everything you need to know - TechTarget"
[2]: https://learn.microsoft.com/en-us/answers/questions/2121835/gpt-4o-has-a-35k-input-token-limit?utm_source=chatgpt.com "GPT-4o has a 35k input token limit - Microsoft Q&A"



In [43]:
loader_summary = UnstructuredMarkdownLoader("data/finance/financial_summary.md", mode="single")
docs_summary  = loader_summary.load()

docs_summary[0].metadata["access_level"] = "finance"
docs_summary[0].metadata["file_name"] = "financial_summary.md"
docs_summary[0].metadata["page_no"] = 1

In [44]:
processed_docs.append(docs_summary[0])
print('After adding finance summary', len(processed_docs))

After adding finance summary 101


In [45]:
# Read the file content
count_tokens_in_file('data/finance/quarterly_financial_report.md')

Total tokens in 'data/finance/quarterly_financial_report.md': 2281


In [46]:
loader_report = UnstructuredMarkdownLoader("data/finance/quarterly_financial_report.md", mode="single")
docs_report  = loader_report.load()
docs_report[0].metadata["access_level"] = "finance"
docs_report[0].metadata["file_name"] = "quarterly_financial_report.md"
docs_report[0].metadata["page_no"] = 1
processed_docs.append(docs_report[0])
print('After adding finance report', len(processed_docs))

After adding finance report 102


#### MARKETING DATASET

In [47]:
loader_q1 = UnstructuredMarkdownLoader("data/marketing/marketing_report_q1_2024.md", mode="single")
docs_q1  = loader_q1.load()
docs_q1[0].metadata["access_level"] = "marketing"
docs_q1[0].metadata["file_name"] = "marketing_report_q1_2024.md"
docs_q1[0].metadata["page_no"] = 1
processed_docs.append(docs_q1[0])

loader_q2 = UnstructuredMarkdownLoader("data/marketing/marketing_report_q2_2024.md", mode="single")
docs_q2  = loader_q2.load()
docs_q2[0].metadata["access_level"] = "marketing"
docs_q2[0].metadata["file_name"] = "marketing_report_q2_2024.md"
docs_q2[0].metadata["page_no"] = 1
processed_docs.append(docs_q2[0])

loader_q3 = UnstructuredMarkdownLoader("data/marketing/marketing_report_q3_2024.md", mode="single")
docs_q3  = loader_q3.load()
docs_q3[0].metadata["access_level"] = "marketing"
docs_q3[0].metadata["file_name"] = "marketing_report_q3_2024.md"
docs_q3[0].metadata["page_no"] = 1
processed_docs.append(docs_q3[0])

loader_q4 = UnstructuredMarkdownLoader("data/marketing/market_report_q4_2024.md", mode="single")
docs_q4  = loader_q4.load()
docs_q4[0].metadata["access_level"] = "marketing"
docs_q4[0].metadata["file_name"] = "market_report_q4_2024.md"
docs_q4[0].metadata["page_no"] = 1
processed_docs.append(docs_q4[0])

loader_q_all = UnstructuredMarkdownLoader("data/marketing/marketing_report_2024.md", mode="single")
docs_q_all  = loader_q_all.load()
docs_q_all[0].metadata["access_level"] = "marketing"
docs_q_all[0].metadata["file_name"] = "marketing_report_2024.md"
docs_q_all[0].metadata["page_no"] = 1
processed_docs.append(docs_q_all[0])

print('After adding marketing dataset', len(processed_docs))

After adding marketing dataset 107


In [48]:
docs_q1[0].metadata['source']

'data/marketing/marketing_report_q1_2024.md'

In [49]:
# Count tokens in page_content
my_list = [docs_q1,docs_q2,docs_q3,docs_q4,docs_q_all]
for i, doc in enumerate(my_list):
    count_tokens_in_file(doc[0].metadata['source'])

Total tokens in 'data/marketing/marketing_report_q1_2024.md': 1269
Total tokens in 'data/marketing/marketing_report_q2_2024.md': 1311
Total tokens in 'data/marketing/marketing_report_q3_2024.md': 1252
Total tokens in 'data/marketing/market_report_q4_2024.md': 1351
Total tokens in 'data/marketing/marketing_report_2024.md': 1393


#### Engineering

In [50]:
count_tokens_in_file('data/engineering/engineering_master_doc.md')

Total tokens in 'data/engineering/engineering_master_doc.md': 6658


In [57]:
from langchain.docstore.document import Document
import os

def parse_markdown_with_headings(file_path, access_level="engineer"):
    with open(file_path, "r", encoding="utf-8") as f:
        lines = f.readlines()

    docs = []
    current_heading = ""
    current_level = ""
    buffer = []
    page_cnt = 1
    file_name = os.path.basename(file_path)

    def flush_buffer(page_cnt, heading, level):
        if heading and buffer:
            content = heading + "\n" + "\n".join(buffer).strip()
            docs.append(Document(
                page_content=content,
                metadata={
                    "file_name": file_name,
                    "page_no": page_cnt,
                    "access_level": access_level,
                    "heading_level": level
                }
            ))
            return page_cnt + 1
        return page_cnt

    for line in lines:
        stripped = line.strip()

        if stripped.startswith("# "):  # Main title (H1)
            page_cnt = flush_buffer(page_cnt, current_heading, current_level)
            current_heading = stripped
            current_level = "H1"
            buffer = []
        elif stripped.startswith("## "):  # Heading (H2)
            page_cnt = flush_buffer(page_cnt, current_heading, current_level)
            current_heading = stripped
            current_level = "H2"
            buffer = []
        elif stripped.startswith("### "):  # Subheading (H3)
            page_cnt = flush_buffer(page_cnt, current_heading, current_level)
            current_heading = stripped
            current_level = "H3"
            buffer = []
        else:
            if stripped:
                buffer.append(stripped)

    # Final flush
    page_cnt = flush_buffer(page_cnt, current_heading, current_level)

    return docs

docs_engineering = parse_markdown_with_headings(file_path = "data/engineering/engineering_master_doc.md",access_level="engineer")

In [58]:
# Example output
for doc in docs_engineering[20:25]:
    print("---")
    print("Content:", doc.page_content)
    print("Metadata:", doc.metadata)

---
Content: ### 6.3 Defect Management
#### 6.3.1 Bug Tracking
* Jira for logging and tracking defects
* Required fields: steps to reproduce, expected vs. actual results, environment
* Severity classification:
* S1 (Critical): System unusable, data corruption, security vulnerability
* S2 (Major): Major function impacted, no workaround
* S3 (Minor): Minor impact, workaround available
* S4 (Cosmetic): UI issues, typos, non-functional issues
#### 6.3.2 Defect SLAs
* S1: Resolution within 24 hours, immediate patch release if needed
* S2: Resolution within 72 hours, included in next scheduled release
* S3: Resolution within 2 weeks
* S4: Prioritized based on business impact
#### 6.3.3 Bug Triage Process
* Daily triage meeting for new bugs
* Weekly bug review for outstanding issues
* Monthly quality metrics review
Metadata: {'file_name': 'engineering_master_doc.md', 'page_no': 21, 'access_level': 'engineer', 'heading_level': 'H3'}
---
Content: ### 7.1 CI/CD Pipeline
#### 7.1.1 Continuous Int

In [59]:
# Append to the processed file
for doc_id in range(len(docs_engineering)):
    processed_docs.append(docs_engineering[doc_id])
print('After adding Engineering report', len(processed_docs))

After adding Engineering report 140


#### General

In [60]:
count_tokens_in_file('data/general/employee_handbook.md')

Total tokens in 'data/general/employee_handbook.md': 3019


In [61]:
docs_general = parse_markdown_with_headings(file_path = "data/general/employee_handbook.md",access_level="employee")

In [62]:
# Example output
for doc in docs_general[:5]:
    print("Content:", doc.page_content)
    print("Metadata:", doc.metadata)
    print("***"*8)

Content: ## Table of Contents
1. [Welcome & Introduction](#welcome--introduction)
2. [Employee Onboarding & Benefits](#employee-onboarding--benefits)
3. [Leave Policies](#leave-policies)
4. [Work Hours & Attendance](#work-hours--attendance)
5. [Code of Conduct & Workplace Behavior](#code-of-conduct--workplace-behavior)
6. [Health & Safety](#health--safety)
7. [Compensation & Payroll](#compensation--payroll)
8. [Reimbursement Policies](#reimbursement-policies)
9. [Training & Development](#training--development)
10. [Performance & Feedback](#performance--feedback)
11. [Privacy & Data Security](#privacy--data-security)
12. [Exit Policy](#exit-policy)
13. [FAQs](#faqs)
14. [Miscellaneous](#miscellaneous)
---
Metadata: {'file_name': 'employee_handbook.md', 'page_no': 1, 'access_level': 'employee', 'heading_level': 'H2'}
************************
Content: ### Company Vision and Mission
At FinSolve Technologies, our vision is to empower businesses and individuals through innovative technology 

In [63]:
# Append to processed_docs
for doc_id in range(len(docs_general)):
    processed_docs.append(docs_general[doc_id])
print('After adding docs_general report', len(processed_docs))

After adding docs_general report 195


#### Index Document Chunks and Embeddings in Vector DB

1. Generate embeddings for each chunk using a model (e.g., OpenAI).

2. Store chunks + embeddings in Chroma for efficient vector similarity search during retrieval.

!pip install langchain-chroma==0.1.4

In [64]:
from langchain_chroma import Chroma

In [65]:
# create vector DB of docs and embeddings
collection_info = "rbac_rag_store"
chroma_db = Chroma.from_documents(documents=processed_docs,
                                  collection_name=collection_info,
                                  embedding=openai_embed_model,
                                  # need to set the distance function to cosine else it uses euclidean by default
                                  # check https://docs.trychroma.com/guides#changing-the-distance-function
                                  collection_metadata={"hnsw:space": "cosine"},
                                  persist_directory="my_db5")

### Load Vector DB from disk

Once a vector database has been created and persisted to disk, you can easily reload it at any time without regenerating embeddings.

- **Cost-Effective**: Generating embeddings (especially using OpenAI or proprietary models) incurs costs. By saving the vector store after initial creation, you avoid repeating that expense
- **Faster Startup**: Loading an existing vector DB from disk is significantly faster than reprocessing and re-embedding large document sets.

In [70]:
# load from disk
collection_info = "rbac_rag_store"
chroma_db = Chroma(persist_directory="my_db5",
                   collection_name = collection_info,
                   embedding_function=openai_embed_model)

print(f"Number of documents in collection: {chroma_db._collection.count()}")


Number of documents in collection: 195


### Semantic Similarity based Retrieval

We use simple cosine similarity here and retrieve the top 3 similar documents based on the user input query

In [71]:
similarity_retriever = chroma_db.as_retriever(search_type="similarity_score_threshold",
                                              search_kwargs={"k": 3, "score_threshold": 0.1})

In [72]:
from IPython.display import display, Markdown

def display_docs(docs):
    for doc in docs:
        print('Metadata:', doc.metadata)
        print('Content Brief:')
        display(Markdown(doc.page_content))
        print()

In [73]:
query = "What is the companys policy on attendance and late arrivals?"
top_docs = similarity_retriever.invoke(query)
display_docs(top_docs)

Metadata: {'access_level': 'employee', 'file_name': 'employee_handbook.md', 'heading_level': 'H3', 'page_no': 11}
Content Brief:


### Attendance & Punctuality
- Attendance must be marked daily via biometric, swipe card, or HRMS app.
- Late arrivals/early departures must be notified to the manager.
- Three or more late arrivals in a month may attract disciplinary action.


Metadata: {'access_level': 'employee', 'file_name': 'employee_handbook.md', 'heading_level': 'H2', 'page_no': 51}
Content Brief:


## FAQs
**Q: How do I apply for maternity leave?**
A: Submit application on HRMS with medical certificate at least 8 weeks before due date.
**Q: What if I forget to mark attendance?**
A: Inform HR within 24 hours; repeated lapses may lead to deduction or disciplinary action.
**Q: How is overtime calculated?**
A: Paid at double the regular wage for hours beyond statutory limits, subject to manager approval.
**Q: Can I work from home?**
A: Up to 2 days/week, subject to manager and business approval.
---


Metadata: {'access_level': 'employee', 'file_name': 'employee_handbook.md', 'heading_level': 'H3', 'page_no': 55}
Content Brief:


### Dress Code & Office Etiquette
- Maintain cleanliness of workspace.
- Use meeting rooms for calls/discussions.
- Mobile phones on silent in work areas.
---
**Note**: This handbook is a living document and may be updated periodically. For policy clarifications or state-specific rules, please consult HR or refer to the company’s policy portal.

## Build the RAG

In [74]:
from langchain_core.prompts import ChatPromptTemplate

rag_prompt = """You are an assistant who is an expert in question-answering tasks.
                Answer the following question using only the following pieces of retrieved context.
                If the answer is not in the context, do not make up answers, just say that you don't know.
                Keep the answer to the point based on the information from the context.

                Question:
                {question}

                Context:
                {context}

                Answer:
            """

# convert the raw template in to structured one,
# so we can easily ingest <question> and <context>
rag_prompt_template = ChatPromptTemplate.from_template(rag_prompt)

In [75]:
from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from langchain_core.runnables import RunnableLambda
from operator import itemgetter


chatgpt = ChatOpenAI(model_name="gpt-4o-mini", temperature=0)

def format_docs(docs):
    return "\n\n".join(doc.page_content for doc in docs)

src_rag_response_chain = (
    {
        "context": (itemgetter('context')
                        |
                    RunnableLambda(format_docs)),
        "question": itemgetter("question")
    }
        |
    rag_prompt_template
        |
    chatgpt
        |
    StrOutputParser()
)

rag_chain_w_sources = (
    {
        "context": similarity_retriever,
        "question": RunnablePassthrough()
    }
        |
    RunnablePassthrough.assign(response=src_rag_response_chain)
)

In [76]:
query = "who will approve my leave request?"
result = rag_chain_w_sources.invoke(query)
result

{'context': [Document(metadata={'access_level': 'employee', 'file_name': 'employee_handbook.md', 'heading_level': 'H3', 'page_no': 8}, page_content='### Leave Application Process\n- Apply via HRMS/leave portal at least 3 days in advance (except emergencies).\n- Sick leave for more than 2 days requires a medical certificate.\n- Emergency leave to be communicated to manager/HR as soon as possible.\n- Approval from reporting manager and HR required for all leave requests.\n- Leave balance can be viewed on the HRMS portal.'),
  Document(metadata={'access_level': 'employee', 'file_name': 'employee_handbook.md', 'heading_level': 'H3', 'page_no': 7}, page_content='### Types of Leave\n| Leave Type | Entitlement/Details |\n|------------|---------------------|\n| **Privilege/Annual** | 15-21 days/year (as per state Shops & Establishments Act); accrued monthly |\n| **Sick Leave** | 12 days/year (non-cumulative; medical certificate for >2 days) |\n| **Casual Leave** | 7 days/year (state-specific) 

## Build the RAG Pipeline

In [79]:
from langchain_core.runnables import RunnableLambda
from operator import itemgetter

chatgpt = ChatOpenAI(model_name="gpt-4o-mini", temperature=0)

# load from disk
collection_info = "rbac_rag_store"
chroma_db = Chroma(persist_directory="my_db5",
                   collection_name = collection_info,
                   embedding_function=openai_embed_model)

print(f"Number of documents in collection: {chroma_db._collection.count()}")

def format_docs(docs):
    return "\n\n".join(doc.page_content for doc in docs)

def get_rbac_retriever(user_role: str):
    """
    Creates a retriever that filters documents based on the user's role.
    """
    return chroma_db.as_retriever(
        search_type="similarity",
        search_kwargs={"k": 3, "filter": {"access_level": user_role}}
    )

src_rag_response_chain = (
    {
        "context": (itemgetter('context')
                        |
                    RunnableLambda(format_docs)),
        "question": itemgetter("question")
    }
        |
    rag_prompt_template
        |
    chatgpt
        |
    StrOutputParser()
)

# This chain now expects a dictionary with "question" and "user_role"
rag_chain_with_rbac = (
    {
        "context": lambda x: get_rbac_retriever(x["user_role"]).invoke(x["question"]),
        "question": itemgetter("question"),
        "user_role": itemgetter("user_role") # Pass the role along for clarity if needed
    }
    |
    RunnablePassthrough.assign(response=src_rag_response_chain)
)

# Example invocation for an engineer
hr_query = "What is the purpose of the engineering document?"
result = rag_chain_with_rbac.invoke({"question": hr_query, "user_role": "engineer"})

print(f"Question: {result['question']}\\n")
print(f"Role: {result['user_role']}\\n")
print(f"Answer: {result['response']}\\n")
print("--- Retrieved Context ---")
display_docs(result['context'])

Number of documents in collection: 195
Question: What is the purpose of the engineering document?\n
Role: engineer\n
Answer: The purpose of the engineering document is to outline the technical architecture, development processes, and operational guidelines for FinSolve's product ecosystem, serving as a comprehensive guide for engineering teams, stakeholders, and partners to ensure alignment with FinSolve's mission.\n
--- Retrieved Context ---
Metadata: {'access_level': 'engineer', 'file_name': 'engineering_master_doc.md', 'heading_level': 'H3', 'page_no': 2}
Content Brief:


### 1.2 Purpose
This engineering document outlines the technical architecture, development processes, and operational guidelines for FinSolve's product ecosystem. It serves as a comprehensive guide for engineering teams, stakeholders, and partners to ensure alignment with FinSolve's mission: "To empower financial freedom through secure, scalable, and innovative technology solutions."


Metadata: {'access_level': 'engineer', 'file_name': 'engineering_master_doc.md', 'heading_level': 'H3', 'page_no': 4}
Content Brief:


### 1.4 Document Control
| Version | Date | Author | Changes |
|---------|------|--------|---------|
| 1.0 | 2025-05-01 | Engineering Team | Initial version |
| 1.1 | 2025-05-14 | Tech Architecture Council | Updated diagrams and monitoring section |


Metadata: {'access_level': 'engineer', 'file_name': 'engineering_master_doc.md', 'heading_level': 'H3', 'page_no': 32}
Content Brief:


### 10.2 Reference Documents
| Document | Location | Purpose |
|----------|----------|---------|
| AWS Well-Architected Framework | Internal Wiki | Cloud architecture best practices |
| PCI-DSS Guidelines | Security Portal | Payment security compliance |
| Kubernetes Documentation | kubernetes.io | Container orchestration reference |
| OWASP Security Standards | Internal Wiki | Web application security |
| Data Protection Policy | Legal Repository | Data handling requirements |